File: JXBC-Server/BossComments/apiserver/appbase/controllers/AccountController.php
<?php

namespace app\appbase\controller;

/**
 * @SWG\Tag(
 *   name="Account",
 *   description="账号相关API，包括注册登录，重置密码等",
 *   @SWG\ExternalDocumentation(
 *     description="Find out more",
 *     url="http://swagger.io"
 *   )
 * )
 */ 
class AccountController
{
    /**
     * @SWG\POST(
     *     path="/appbase/Account/createNew",
     *     summary="创建APP匿名账号",
     *     description="使用客户端设备信息，创建匿名账号；<br/>浏览器首次访问网站时(服务器端自动写入cookie)<br/>客户端调用时机：<br/>1.APP首次启动;<br/>2.APP退出账号后;",     
     *     tags={"Account"},
     *     @SWG\Parameter(
     *         name="body",
     *         in="body",
     *         description="客户端设备信息",
     *         required=true,
     *         @SWG\Schema(ref="#/definitions/ClientDevice"),
     *     ),       
     *     @SWG\Response(
     *         response=200,
     *         description="账号信息",
     *         @SWG\Schema(
     *             ref="#/definitions/AccountSignResult"
     *         )
     *     ),
     *     @SWG\Response(
     *         response="412",
     *         description="不符合预期的输入参数",
     *         @SWG\Schema(   
     *             ref="#/definitions/Error"
     *         )
     *     )
     * )
     */
    public function createNew($entity)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI"); 
    }
    
    /**
     * @SWG\Definition(
     *   definition="AccountSign:signUp",
     *   allOf={
     *     @SWG\Schema(ref="#/definitions/AccountSign"),
     *     @SWG\Schema(required={"ValidationCode","SelectedProfileType"})
     *   },
     * )    
     * @SWG\Post(
     *     path="/appbase/Account/signUp",
     *     summary="注册账号",
     *     tags={"Account"},
     *     description="asd",
     *     @SWG\Parameter(
     *         name="body",
     *         in="body",
     *         description="账号注册信息",
     *         required=true,
     *         @SWG\Schema(ref="#/definitions/AccountSign:signUp"),
     *     ),     
     *     @SWG\Response(
     *         response=200,
     *         description="账号信息",
     *         @SWG\Schema(
     *             ref="#/definitions/AccountSignResult"
     *         )
     *     ),
     *     @SWG\Response(
     *         response="412",
     *         description="不符合预期的输入参数",
     *         @SWG\Schema(   
     *             ref="#/definitions/Error"
     *         )
     *     )
     * )
     */
    public function signUp($entity)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI"); 
    }
    
    /**
     * @SWG\Post(
     *     path="/appbase/Account/signIn",
     *     summary="登录（用户名+密码）",
     *     tags={"Account"},
     *     @SWG\Parameter(
     *         name="body",
     *         in="body",
     *         description="登录信息",
     *         required=true,
     *         @SWG\Schema(ref="#/definitions/AccountSign"),
     *     ),     
     *     @SWG\Response(
     *         response=200,
     *         description="账号信息",
     *         @SWG\Schema(
     *             ref="#/definitions/AccountSignResult"
     *         )
     *     ),
     *     @SWG\Response(
     *         response="412",
     *         description="不符合预期的输入参数",
     *         @SWG\Schema(   
     *             ref="#/definitions/Error"
     *         )
     *     )
     * )
     */
    public function signIn($entity)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI"); 
    }

    /**
     * @SWG\Post(
     *     path="/appbase/Account/signInByToken",
     *     summary="登录（TOKEN）",
     *     tags={"Account"},
     *     @SWG\Response(
     *         response=200,
     *         description="账号信息",
     *         @SWG\Schema(
     *             ref="#/definitions/AccountSignResult"
     *         )
     *     ),
     *     @SWG\Response(
     *         response="412",
     *         description="不符合预期的输入参数",
     *         @SWG\Schema(
     *             ref="#/definitions/Error"
     *         )
     *     )
     * )
     */
    public function signInByToken()
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI");
    }
    
    /**
     * @SWG\Post(
     *     path="/appbase/Account/sendValidationCode",
     *     summary="发送手机验证码",
     *     tags={"Account"},
     *     consumes={"text/plain"},
     *     @SWG\Parameter(
     *         name="phone",
     *         in="query",
     *         description="手机号",
     *         required=true,
     *         type="string"
     *     ),     
     *     @SWG\Response(
     *         response=200,
     *         description="发送是否成功",
     *         @SWG\Schema(type="boolean")     
     *     )
     * )
     */
    public function sendValidationCode($phone)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI"); 
    }  

    /**
     * @SWG\Get(
     *     path="/appbase/Account/checkValidationCode",
     *     summary="验证手机验证码是否正确",
     *     tags={"Account"},
     *     consumes={"text/plain"},
     *     @SWG\Parameter(
     *         name="phone",
     *         in="query",
     *         description="手机号",
     *         required=true,
     *         type="string"
     *     ), 
     *     @SWG\Parameter(
     *         name="code",
     *         in="query",
     *         description="验证码",
     *         required=true,
     *         type="string"
     *     ), 
     *     @SWG\Response(
     *         response=200,
     *         description="",
     *         @SWG\Schema(type="boolean")     
     *     )
     * )
     */
    public function checkValidationCode($phone, $code)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI"); 
    } 
    
    /**
     * @SWG\Get(
     *     path="/appbase/Account/existsMobilePhone",
     *     summary="验证手机号是否注册",
     *     tags={"Account"},
     *     consumes={"text/plain"},
     *     @SWG\Parameter(
     *         name="phone",
     *         in="query",
     *         description="手机号",
     *         required=true,
     *         type="string"
     *     ),     
     *     @SWG\Response(
     *         response=200,
     *         description="",
     *         @SWG\Schema(type="boolean")     
     *     )
     * )
     */
    public function existsMobilePhone($phone)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI"); 
    }

    /**
     * @SWG\Definition(
     *   definition="AccountSign:reset",
     *   allOf={
     *     @SWG\Schema(ref="#/definitions/AccountSign"),
     *     @SWG\Schema(required={"ValidationCode"})
     *   },
     * )
     * @SWG\Post(
     *     path="/appbase/Account/resetPassword",
     *     summary="重置密码",
     *     tags={"Account"},
     *     @SWG\Parameter(
     *         name="body",
     *         in="body",
     *         description="用户名+密码+手机验证码",
     *         required=true,
     *         @SWG\Schema(ref="#/definitions/AccountSign:reset"),
     *     ),
     *     @SWG\Response(
     *         response=200,
     *         description="中直接过",
     *         @SWG\Schema(
     *             ref="#/definitions/AccountSignResult"
     *         )
     *     )
     * )
     */
    public function resetPassword($entity)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI");
    }
}

File: JXBC-Server/appbase/src/JXBC.Web/JXBC.WebAPI/Controllers/AccountController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Web.Http;
using JXBC.Passports;
using JXBC.WebCore;
using JXBC.WebCore.Mvc;
using JXBC.WebCore.ViewModels;
using JXBC.Workplace;
using JXBC.Passports.Security;
using M2SA.AppGenome.Logging;

namespace JXBC.WebAPI.Controllers
{
    /// <summary>
    /// 
    /// </summary>
    public class AccountController : ApiController
    {
        private void WriteTokenToBrowser(AccountSignResult signResult)
        {
            if (null == signResult || signResult.SignStatus != SignStatus.Success)
                return;

            if (null != this.Request.Headers.UserAgent)
            {
                if(signResult.Account.PassportId > 0)
                    CookieHelper.SetCookie(AccountAuthentication.TokenKey, signResult.Account.Token.AccessToken);
                else 
                    CookieHelper.SetCookie(AccountAuthentication.TokenKey, signResult.Account.Token.AccessToken, DateTime.MaxValue);
            }
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="entity"></param>
        /// <returns></returns>
        [HttpPost]
        public AccountSignResult CreateNew(ClientDevice entity)
        {
            var result = new AccountSignResult();
            result.SignStatus = SignStatus.None;

            if (null == entity
                || string.IsNullOrEmpty(entity.DeviceKey)
                || string.IsNullOrEmpty(entity.Product)
                || string.IsNullOrEmpty(entity.Brand))
            {
                this.ReturnPreconditionFailedMessage();
            }

            var account = AccountAuthentication.CreateNew(entity);
            if(null != account)
            {
                result.SignStatus = SignStatus.Success;
                result.Account = new AccountEntity(account, null);

                WriteTokenToBrowser(result);
            }
            else
            {
                result.SignStatus = SignStatus.Error;
            }
                
            return result;
        }

        /// <summary>
        /// 账户注册
        /// </summary>
        /// <param name="entity">必填项：MobilePhone;ValidationCode;EntName;LegalRepresentative;</param>
        /// <returns></returns>
        [HttpPost]
        public AccountSignResult SignUp(AccountSign entity)
        {
            if (null == entity || string.IsNullOrEmpty(entity.MobilePhone) || string.IsNullOrEmpty(entity.Password)
                || string.IsNullOrEmpty(entity.ValidationCode) || ProfileType.None == entity.SelectedProfileType) this.ReturnPreconditionFailedMessage();
            var account = MvcContext.Current.ClientAccount;
            if (null == account) this.ReturnPreconditionFailedMessage();

            var isValid = MessageHelper.CheckSMSValidationCode(entity.MobilePhone, entity.ValidationCode);
            if (MvcContext.Current.Test && entity.ValidationCode == AppEnvironment.TestValidationCode) isValid = true;
            if (false == isValid)
            {
                return new AccountSignResult()
                {
                    SignStatus = SignStatus.InvalidValidationCode,
                    ErrorMessage = "验证码无效，请重新获取"
                };
            }
            if (!string.IsNullOrEmpty(entity.InviteCode))
            {
                System.Web.HttpContext.Current.Items.Add(WorkplaceApplication.InviteCodeKey, entity.InviteCode);
            }
            
            var signStatus = SignStatus.Error;

            var signedUpInfo = new SignedUpInfo()
            {
                InviteCode = entity.InviteCode,
                SignedUpIp = this.Request.GetClientIP(),
                HttpUserAgent = this.Request.GetUserAgent(),
                HttpReferer = null == this.Request.Headers.Referrer ? "" : this.Request.Headers.Referrer.ToString()
            };
            var userPassport = AccountAuthentication.SignUp(entity.MobilePhone, entity.Password, entity.SelectedProfileType, signedUpInfo, out signStatus);

            var result = new AccountSignResult();
            result.SignStatus = signStatus;
            if (result.SignStatus == SignStatus.Success)
            {
                result.CreatedNewPassport = true;
                result.Account = new AccountEntity(account, userPassport);
                result.AdditionalAction = ProcessAdditionalAction(entity.AdditionalAction);

                WriteTokenToBrowser(result);
            }
            else
            {
                if (result.SignStatus == SignStatus.DuplicateMobilePhone)
                    result.ErrorMessage = "手机号已经注册";
                else
                    result.ErrorMessage = "注册失败";
            }

            return result;
        }

        [HttpPost]
        public AccountSignResult SignIn(AccountSign entity)
        {
            if (null == entity || string.IsNullOrEmpty(entity.MobilePhone) || string.IsNullOrEmpty(entity.Password)) return null;
            var account = MvcContext.Current.ClientAccount;
            if (null == account) return null;

            if (!string.IsNullOrEmpty(entity.InviteCode))
            {
                System.Web.HttpContext.Current.Items.Add(WorkplaceApplication.InviteCodeKey, entity.InviteCode);
            }
            UserPassport userPassport = null;
            var isSignIn = AccountAuthentication.SignIn(entity.MobilePhone, entity.Password, new SignedInLog(), out userPassport);

            var result = new AccountSignResult();
            result.SignStatus = isSignIn ? SignStatus.Success : SignStatus.InvalidPassword;
            if (result.SignStatus == SignStatus.Success)
            {
                result.Account = new AccountEntity(account, userPassport);
                result.AdditionalAction = ProcessAdditionalAction(entity.AdditionalAction);

                WriteTokenToBrowser(result);
            }
            else
            {
                result.ErrorMessage = "用户名或密码错误";
            }
            return result;
        }

        [HttpPost]
        public AccountSignResult ShortcutSignIn(AccountSign entity)
        {
            if (null == entity || string.IsNullOrEmpty(entity.MobilePhone) || string.IsNullOrEmpty(entity.ValidationCode))
                return null;
            var account = MvcContext.Current.ClientAccount;
            if (null == account) return null;

            var passportId = UserPassport.FindIdByMobilePhone(entity.MobilePhone);
            if(passportId == 0)
            {
                entity.Password = HashHelper.ComputeHash(entity.MobilePhone, HashAlgorithmName.SHA1).Substring(0, 6);
                return SignUp(entity);
            }

            var isValid = MessageHelper.CheckSMSValidationCode(entity.MobilePhone, entity.ValidationCode);
            if (MvcContext.Current.Test && entity.ValidationCode == AppEnvironment.TestValidationCode) isValid = true;
            if (false == isValid)
            {
                return new AccountSignResult()
                {
                    SignStatus = SignStatus.InvalidValidationCode,
                    ErrorMessage = "验证码无效，请重新获取"
                };
            }
            if (!string.IsNullOrEmpty(entity.InviteCode))
            {
                System.Web.HttpContext.Current.Items.Add(WorkplaceApplication.InviteCodeKey, entity.InviteCode);
            }
            UserPassport userPassport = null;
            var isSignIn = AccountAuthentication.SignIn(passportId, new SignedInLog(), out userPassport);

            var result = new AccountSignResult();
            result.SignStatus = isSignIn ? SignStatus.Success : SignStatus.InvalidPassword;
            if (result.SignStatus == SignStatus.Success)
            {
                result.Account = new AccountEntity(account, userPassport);
                result.AdditionalAction = ProcessAdditionalAction(entity.AdditionalAction);

                WriteTokenToBrowser(result);
            }
            else
            {
                result.ErrorMessage = "用户名或密码错误";
            }
            return result;
        }

        [HttpPost]
        public AccountSignResult BindThirdPassport(ThirdPassport entity)
        {
            if (null == entity || string.IsNullOrEmpty(entity.Platform) || string.IsNullOrEmpty(entity.PlatformPassportId)) return null;
            var account = MvcContext.Current.ClientAccount;
            if (null == account) return null;

            var signStatus = SignStatus.Error;
            UserPassport userPassport = AccountAuthentication.BindThirdPassport(entity, out signStatus);

            var result = new AccountSignResult();
            result.SignStatus = signStatus;
            if (result.SignStatus == SignStatus.Success)
            {
                result.Account = new AccountEntity(account, userPassport);
                WriteTokenToBrowser(result);
            }
            else
            {
                result.ErrorMessage = "绑定账号失败";
            }
            return result;
        }

        [HttpPost]
        public AccountSignResult SignInByToken()
        {
            if (false == MvcContext.Current.IsAuthenticated)
                return new AccountSignResult()
                {
                    SignStatus = SignStatus.Failed,
                    ErrorMessage = "认证信息已失效，请重新登录"
                };

            var account = new AccountEntity(MvcContext.Current.ClientAccount, MvcContext.Current.UserPassport);

            return new AccountSignResult()
            {
                SignStatus = SignStatus.Success,
                Account = account
            };
        }

        [HttpPost]
        public bool SignOut()
        {
            var account = MvcContext.Current.ClientAccount;
            if (null == account) return false;

            return true;
        }

        [HttpPost]
        public AccountSignResult ResetPassword(AccountSign entity)
        {
            if (null == entity || string.IsNullOrEmpty(entity.MobilePhone)
                || string.IsNullOrEmpty(entity.Password) || string.IsNullOrEmpty(entity.ValidationCode)) return null;
            var account = MvcContext.Current.ClientAccount;
            if (null == account) return null;

            var isValid = MessageHelper.CheckSMSValidationCode(entity.MobilePhone, entity.ValidationCode);
            if (MvcContext.Current.Test && entity.ValidationCode == AppEnvironment.TestValidationCode) isValid = true;
            if (false == isValid)
            {
                return new AccountSignResult()
                {
                    SignStatus = SignStatus.InvalidValidationCode,
                    ErrorMessage = "验证码无效，请重新获取"
                };
            }

            var passportId = UserPassport.FindIdByMobilePhone(entity.MobilePhone);
            if (passportId < 1)
            {
                return new AccountSignResult()
                {
                    SignStatus = SignStatus.InvalidMobilePhone,
                    ErrorMessage = "手机号未注册"
                };
            }

            var isChanged = MemberShip.ChangePassword(passportId, entity.Password);
            if (isChanged)
            {
                return new AccountSignResult()
                {
                    SignStatus = SignStatus.Success
                };
            }
            return new AccountSignResult()
            {
                SignStatus = SignStatus.Failed
            };
        }

        [HttpPost]
        public bool SendValidationCode(string phone)
        {
            if (string.IsNullOrEmpty(phone)) return false;
            var account = MvcContext.Current.ClientAccount;
            if (null == account) return false;

            var result = MessageHelper.SendSMSValidationCode(account, this.Request.GetClientIP(), phone);
            return result.Success;
        }

        [HttpGet]
        public bool CheckValidationCode(string phone, string code)
        {
            var isValid = MessageHelper.CheckSMSValidationCode(phone, code);
            if (MvcContext.Current.Test && code == AppEnvironment.TestValidationCode) isValid = true;
            return isValid;
        }

        [HttpGet]
        public bool ExistsMobilePhone(string phone)
        {
            if (string.IsNullOrEmpty(phone)) return false;
            var passportId = UserPassport.FindIdByMobilePhone(phone);
            return passportId > 0;
        }


        private AdditionalAction ProcessAdditionalAction(AdditionalAction action)
        {
            //Todo: ProcessAdditionalAction
            return action;
        }
    }
}
